torch/_tensor.py

```diff
@@ -103,7 +110,24 @@ def __reduce_ex__(self, proto):
         if has_torch_function_unary(self):
             return handle_torch_function(Tensor.__reduce_ex__, (self,), self, proto)
         func, args = self._reduce_ex_internal(proto)
-        return (_rebuild_from_type, (func, type(self), args, self.__dict__))
+        # Get the state of the python subclass
+        # This loosely mimicks the function on the object class but since Tensor do not inherit
+        # from it, we cannot call that function directly
+        # https://github.com/python/cpython/blob/c83919bd635f4433f1c6ae8504996a9fe3c215e5/Objects/typeobject.c#L4891
+        if hasattr(self, "__getstate__"):
+            state = self.__getstate__()
+        else:
+            # Contrary to the cpython function, we unify dict and slots into a single dictionary
+            # instead of a tuple of two dictionaries
+            if hasattr(self.__class__, "__slotnames__"):
+                attr_to_save = self.__class__.__slotnames__
+            else:
+                import copyreg
+                attr_to_save = copyreg._slotnames(self.__class__)
```

Comment:
Should we replace this with a manual implementation here?

Reply:
idk. When is slotnames available versus not? Is there a reference?

Reply:
It is not documented. But the CPython code actually uses it.
But I can't find any info on which versions of Python actually have that function in copyreg or not.
Should I just copy-paste that function here?

For reference, the function is here: https://github.com/python/cpython/blob/435a0334d341e5f8faed594d9f015746bb7845db/Lib/copyreg.py#L109-L158

Reply:
Based on what you linked, it seems like you should just always call copyreg

Reply:
Yes. For some reason the cpython code re-checks the caching even though the function is doing it internally. Updated.

I'm not sure what is the criterion for using  the copyreg version vs copying?
If the function is there for all versions of python that we support, we're happy with using the internal API (and we'll fix it in future version if it gets removed)?

Reply:
yeah i'd just do it this way with copyreg

Reply:
From looking at the history, the function has been here for 19 years and the only change in the past 14 years has been a minor bug-fix 4 years ago.
So I'm going to assume this is a pretty stable function and just use it :D